File: bak/TextReader.cs
using System;
using UnityEngine;

public class TextReader
{
    private readonly string m_text;
    private int m_position;

    public int Position
    {
        get => m_position;
        private set => m_position = Math.Max(0, value);
    }

    public TextReader(string text)
    {
        m_text = text;
        Position = 0;
    }

    public char Read()
    {
        if (Position < m_text.Length)
        {
            var token = m_text[Position];
            Position++;
            return token;
        }
        return default;
    }

    public char Peak()
    {
        if (Position < m_text.Length)
        {
            var token = m_text[Position];
            return token;
        }
        return default;
    }

    public void Seek(int pos)
    {
        Position = pos;
    }

    public void Rewind(int offset = 1)
    {
        if (offset < 0) throw new Exception("offset must > 0");
        Position -= offset;
    }

    public void Advance(int offset)
    {
        offset = Mathf.Max(0, offset);
        Position += offset;
    }

    public bool CanRead()
    {
        return Position < m_text.Length;
    }

    public void Jump(char c)
    {
        var pos = Position;
        while (CanRead())
        {
            var tok = Read();
            if (tok == c)
            {
                return;
            }
        }
        Seek(pos);
    }

    public void Jump(string c)
    {
        if (!CanRead())
        {
            return;
        }

        var ti = m_text.IndexOf(c, Position);
        if (ti < 0)
        {
            Seek(m_text.Length + 1);
        }
        else
        {
            Seek(ti);
        }
    }

    public bool JumpBack(string c)
    {
        var ti = m_text.LastIndexOf(c, Position);
        if (ti < 0)
        {
            return false;
        }
        else
        {
            Seek(ti);
            return true;
        }
    }

    public void JumpBraceBracketEnd()
    {
        Debug.Assert(Read() == '{');
        var depth = 1;

        while (CanRead())
        {
            var tok = Read();
            if (tok == '{')
            {
                depth++;
            }
            else if (tok == '}')
            {
                depth--;
                if (depth == 0)
                {
                    return;
                }
            }
        }

        throw new Exception("");
    }

    public string PeekString()
    {
        var str = m_text.Substring(0, Position);
        return str;
    }

    public string PeekString(int start, int end)
    {
        var str = m_text.Substring(start, end - start);
        return str;
    }

    public void ForwardToSentenceEnd()
    {
        //while (CanRead())
        //{
        //    var tok = Read();
        //    if (tok.Type == TokenType.SemiColon)
        //    {
        //        return;
        //    }
        //}
    }
}

File: src/tool/TextReader.ts
import assert from "assert";

class TextReader {
    private readonly m_text: string;
    private m_position: number;

    constructor(text: string) {
        this.m_text = text;
        this.m_position = 0;
    }

    get Position(): number {
        return this.m_position;
    }

    set Position(position: number) {
        this.m_position = Math.max(0, position);
    }

    public read(): string | null {
        if (this.m_position >= this.m_text.length) {
            return null;
        }

        const character = this.m_text.charAt(this.m_position);
        this.m_position++;

        return character;
    }

    public Peak(): string | null {
        if (this.m_position < this.m_text.length) {
            const character = this.m_text.charAt(this.m_position);
            return character;
        }
        return null;
    }

    public Seek(pos: number): void {
        this.Position = pos;
    }

    public Rewind(offset: number = 1): void {
        if (offset < 0) {
            throw new Error("offset must > 0");
        }
        this.Position -= offset;
    }

    public Advance(offset: number): void {
        offset = Math.max(0, offset);
        this.Position += offset;
    }

    public CanRead(): boolean {
        return this.Position < this.m_text.length;
    }

    public JumpToChar(c: string): void {
        const pos = this.Position;
        while (this.CanRead()) {
            const tok = this.read();
            if (tok === c) {
                return;
            }
        }
        this.Seek(pos);
    }

    public JumpToString(c: string): void {
        if (!this.CanRead()) { return; }
        const ti = this.m_text.indexOf(c, this.Position);
        if (ti < 0) {
            this.Seek(this.m_text.length + 1);
        } else {
            this.Seek(ti);
        }
    }

    public JumpBackToString(c: string): boolean {
        const ti = this.m_text.lastIndexOf(c, this.Position);
        if (ti < 0) {
            return false;
        } else {
            this.Seek(ti);
            return true;
        }
    }

    public JumpBraceBracketEnd(): void {
        assert(this.read() === "{");
        let depth = 1;

        while (this.CanRead()) {
            const tok = this.read();
            if (tok === "{") {
                depth++;
            } else if (tok === "}") {
                depth--;
                if (depth === 0) {
                    return;
                }
            }
        }
        throw new Error("can not find end brace bracket");
    }

    public PeekString(): string {
        const str = this.m_text.substring(0, this.Position);
        return str;

    }

    public GetString(start: number, end: number): string {
        const str = this.m_text.substring(start, end);
        return str;
    }
}
